File: src/main/go/g0301_0400/s0322_coin_change/solution.go
package s0322_coin_change

// #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Breadth_First_Search
// #Algorithm_II_Day_18_Dynamic_Programming #Dynamic_Programming_I_Day_20
// #Level_2_Day_12_Dynamic_Programming #Top_Interview_150_1D_DP #Big_O_Time_O(m*n)_Space_O(amount)
// #2025_05_10_Time_10_ms_(79.59%)_Space_8.61_MB_(51.21%)

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := range dp {
		dp[i] = amount + 1
	}

	dp[0] = 0

	for i := 1; i <= amount; i++ {
		for _, c := range coins {
			if i < c {
				continue
			}

			dp[i] = min(dp[i], dp[i-c]+1)
		}
	}

	if dp[amount] == amount+1 {
		return -1
	} else {
		return dp[amount]
	}
}

File: src/main/swift/g0301_0400/s0322_coin_change/Solution.swift
// #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Breadth_First_Search
// #Algorithm_II_Day_18_Dynamic_Programming #Dynamic_Programming_I_Day_20
// #Level_2_Day_12_Dynamic_Programming #Top_Interview_150_1D_DP #Big_O_Time_O(m*n)_Space_O(amount)
// #2024_07_02_Time_9_ms_(94.72%)_Space_15.5_MB_(61.62%)

class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        if amount < 1 { return 0 }
        let dummyValue: Int = .max
        var minCoinsHolder = Array(repeating: dummyValue, count: amount + 1)
        minCoinsHolder[0] = 0

        for intermediateAmount in 1 ... amount {
            var minCoins = dummyValue

            for coin in coins {
                if intermediateAmount >= coin {
                    let reminder = intermediateAmount - coin
                    let minCoinsForReminder = minCoinsHolder[reminder]
                    if minCoinsForReminder == dummyValue { continue }
                    minCoins = min(minCoins, 1 + minCoinsForReminder)
                }
            }

            minCoinsHolder[intermediateAmount] = minCoins
        }

        return minCoinsHolder[amount] == dummyValue ? -1 : minCoinsHolder[amount]
    }
}
